app/core/SnapKeyring/SnapKeyring.ts

```diff
@@ -1,132 +1,215 @@
-
-      redirectUser: async (snapId: string, url: string, message: string) => {
-        Logger.log(
-          `SnapKeyring: redirectUser called with \n
-        - snapId: ${snapId} \n
-        - url: ${url} \n
-        - message: ${message} \n`,
+export interface SnapKeyringBuilder {
+  (): SnapKeyring;
+  type: typeof SnapKeyring.type;
+}
+
+/**
+ * Helpers for the Snap keyring implementation.
+ */
+export interface SnapKeyringHelpers {
+  persistKeyringHelper: () => Promise<void>;
+  removeAccountHelper: (address: string) => Promise<void>;
+}
+
+class SnapKeyringImpl implements SnapKeyringCallbacks {
+  readonly #messenger: SnapKeyringBuilderMessenger;
+
+  readonly #persistKeyringHelper: SnapKeyringHelpers['persistKeyringHelper'];
+
+  readonly #removeAccountHelper: SnapKeyringHelpers['removeAccountHelper'];
+
+  constructor(
+    messenger: SnapKeyringBuilderMessenger,
+    { persistKeyringHelper, removeAccountHelper }: SnapKeyringHelpers,
+  ) {
+    this.#messenger = messenger;
+    this.#persistKeyringHelper = persistKeyringHelper;
+    this.#removeAccountHelper = removeAccountHelper;
+  }
+
+  async addressExists(address: string) {
+    const addresses = await this.#messenger.call(
+      'KeyringController:getAccounts',
+    );
+    return addresses.includes(address.toLowerCase());
+  }
+
+  async saveState() {
+    await this.#persistKeyringHelper();
+  }
+
+  private async withApprovalFlow<Return>(
+    run: (flowId: string) => Promise<Return>,
+  ): Promise<Return> {
+    const { id: flowId } = this.#messenger.call('ApprovalController:startFlow');
+
+    try {
+      return await run(flowId);
+    } finally {
+      this.#messenger.call('ApprovalController:endFlow', {
+        id: flowId,
+      });
+    }
+  }
+
+  private async addAccountConfirmations({
+    snapId,
+    handleUserInput,
+    accountNameSuggestion,
+  }: {
+    snapId: SnapId;
+    accountNameSuggestion: string;
+    handleUserInput: (accepted: boolean) => Promise<void>;
+  }): Promise<{ accountName?: string }> {
+    return await this.withApprovalFlow(async (_) => {
+      // Show the account **renaming** confirmation dialog.
+      const { success, name: accountName } =
+        await showAccountNameSuggestionDialog(
+          snapId,
+          this.#messenger,
+          accountNameSuggestion,
         );
-      },
-
-      saveState: async () => {
-        await persistKeyringHelper();
-      },
-
-      addAccount: async (
-        address: string,
-        snapId: string,
-        handleUserInput: (accepted: boolean) => Promise<void>,
-        accountNameSuggestion = '',
-      ) => {
-        const { id: addAccountFlowId } = controllerMessenger.call(
-          'ApprovalController:startFlow',
+
+      if (!success) {
+        // User has cancelled account creation
+        await handleUserInput(success);
+
+        throw new Error('User denied account creation');
+      }
+
+      await handleUserInput(success);
```

Comment:
I've missed that when implementing it on the extension, and @danroc changed it recently when adding the new `skipAccountNameSuggestion` flag:
- See: https://github.com/MetaMask/metamask-extension/blob/main/app/scripts/lib/snap-keyring/snap-keyring.ts#L278-L282

```suggestion
      await handleUserInput(success);

      if (!success) {
        throw new Error('User denied account creation');
      }
```

Reply:
I have modified this to wok the same way as extension and allow for skipping the name suggestion modal.